### PR TITLE
Add `Python` package module for Python+Chapel interop

### PR DESCRIPTION
**Pull request by jabraham17**
Adds a new package module, `Python`, which supports calling Python code from Chapel.

For example, the following Chapel progam use the python interface to `BeautifulSoup` to parse HTML

```
import Python;
use URL;
use List;

proc main() {

  const url = 'https://chapel-lang.org/docs/main/language/spec/interoperability.html';
  var htmlReader = openUrlReader(url);
  var html = htmlReader.readAll(string);

  var interp = new Python.Interpreter();
  var mod = new Python.Module(interp, "bs4");

  var cls = new Python.Class(mod, "BeautifulSoup");
  var soup = cls(html, 'html.parser');

  var res: list(owned Python.ClassInstance?);
  res = soup.callMethod(res.type, "find_all", "h3");
  for c in res {
    writeln(c!.getAttr(string, "text"));
  }
}
```

As a another example, this simple program compiles an runs a Python lambda from Chapel code
```chapel
import Python;

config const n = 10;
config const func = "lambda x,: x + 1 if x % 2 != 0 else x";

proc apply(interp: borrowed, type t, arr, l) {
  var lambdaFunc = new Python.Function(interp, l);
  var res: [arr.domain] t;
  for i in arr.domain {
    res(i) = lambdaFunc(t, arr(i));
  }
  return res;
}
proc main() {
  var interp = new Python.Interpreter();

  var data: [1..#n] int = 1..#n;
  writeln(" data: ", data);

  var res = apply(interp, int, data, func);
  writeln("res: ", res);
}

```

### CODE REVIEW

File: modules/packages/Python.chpl [hunk 1]
```diff
@@ -0,0 +1,1417 @@
...
+      Compile a lambda function like 'lambda x,: x + 1'
+
+      Note: this only works with lambdas that a tuple of arguments, like 'x,' or '*args'
+    */
+    @chpldoc.nodoc
+    proc compileLambda(l: string): owned PyObject throws {
+      var globals = chpl_PyEval_GetFrameGlobals();
+      // create a globals if it doesn't exist
+      if !globals {
+        globals = Py_BuildValue("{}"); // TODO: need to decrement this?
```
Comment (DanilaFe) [inline modules/packages/Python.chpl:340]: you do need to decref this. Normally, Python increfs automatically for things that return objects, including `GetFrameGlobals`, so I'd expect this to be safe. If not, set a bool flag if you use `Py_BuildValue`, and decref then.

File: modules/packages/Python.chpl [hunk 2]
```diff
@@ -0,0 +1,1417 @@
...
+
+      .. note::
+
+         This method should be called after any Python API call that may fail.
+         The wrapping types in this module will call this method automatically,
+         most users should not need to call this method directly.
+    */
+    inline proc checkException() throws {
+      var ptype, pvalue, ptraceback: c_ptr(void);
+      PyErr_Fetch(c_ptrTo(ptype), c_ptrTo(pvalue), c_ptrTo(ptraceback));
```
Comment (DanilaFe) [inline modules/packages/Python.chpl:362]: A little birdie tells me that `PyErr_Fetch` is deprecated and you should use [`GetRaisedException`](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_GetRaisedException) instead.

File: modules/packages/Python.chpl [hunk 3]
```diff
@@ -0,0 +1,1417 @@
...
+
+      if pvalue {
+        var pvalue_str = PyObject_Str(pvalue);
+        var value_str = this.fromPython(string, pvalue_str);
+        Py_DECREF(pvalue_str);
+        throw new PythonException(value_str);
```
Comment (DanilaFe) [inline modules/packages/Python.chpl:368]: Since exceptions can be caught, I think you need to make sure to `DECREF` `pvalue`, `ptype` and `ptraceback`. Does `defer` get invoked if we're unwinding?

File: modules/packages/Python.chpl [hunk 4]
```diff
@@ -0,0 +1,1417 @@
...
+         This method should be called after any Python API call that may fail.
+         The wrapping types in this module will call this method automatically,
+         most users should not need to call this method directly.
+    */
+    inline proc checkException() throws {
+      var ptype, pvalue, ptraceback: c_ptr(void);
+      PyErr_Fetch(c_ptrTo(ptype), c_ptrTo(pvalue), c_ptrTo(ptraceback));
+
+      if pvalue {
+        var pvalue_str = PyObject_Str(pvalue);
```
Comment (DanilaFe) [inline modules/packages/Python.chpl:365]: It would help me (and anyone reading this code) to define type aliases for `void*` corresponding to different types. E.g.,

```
type PyObjectPtr = c_ptr(void);
```

And explicitly annotate types here and in the prototype declarations of these functions. This way, we have at least a notional idea of what the types are.

File: modules/packages/Python.chpl [hunk 5]
```diff
@@ -0,0 +1,1417 @@
...
+         a :type:`TypeConverter`.
+    */
+    proc toPython(const val: ?t): c_ptr(void) throws {
+      for converter in this.converters {
+        if converter.handlesType(t) {
+          return converter.toPython(this, t, val);
+        }
+      }
+      // if no converter is found, use the defaults
+      if isIntType(t) {
```
Comment (DanilaFe) [inline modules/packages/Python.chpl:387]: can this be a `case` statement?

File: modules/packages/Python.chpl [hunk 6]
```diff
@@ -0,0 +1,1417 @@
...
+    /*
+      Converts a python list to an array
+    */
+    @chpldoc.nodoc
+    proc fromList(type T, obj: c_ptr(void)): T throws
+      where isArrayType(T) {
+
+      if (PySequence_Check(obj)) {
+        // if its a sequence with a known size, we can just iterate over it
+        var res: T;
```
Comment (DanilaFe) [inline modules/packages/Python.chpl:523]: should this be `noinit` so that you don't need a default-initializer for `T.eltType`?

File: modules/packages/Python.chpl [hunk 7]
```diff
@@ -0,0 +1,1417 @@
...
+    }
+
+    /*
+      Converts a python list to an array
+    */
+    @chpldoc.nodoc
+    proc fromList(type T, obj: c_ptr(void)): T throws
+      where isArrayType(T) {
+
+      if (PySequence_Check(obj)) {
```
Comment (DanilaFe) [inline modules/packages/Python.chpl:521]: Consider flattening this as:

```Chapel
if !PySequence_Check(obj) {
  throw ...
}

// the rest...
```

The large bodies of the if/else make it hard to follow.

File: modules/packages/Python.chpl [hunk 8]
```diff
@@ -0,0 +1,1417 @@
...
+  class Function {
+    @chpldoc.nodoc
+    var fnName: string;
+    @chpldoc.nodoc
+    var fn: owned PyObject?;
+    proc init(mod: borrowed Module, in fnName: string) throws {
+      this.fnName = fnName;
+      this.fn = new PyObject(mod.mod.interpreter, PyObject_GetAttrString(mod.mod.get(), fnName.c_str()));
+      init this;
+      this.fn!.check();
```
Comment (DanilaFe) [inline modules/packages/Python.chpl:760]: Consider invoking the name/object constructor (init#2) to share some code.

File: modules/packages/Python.chpl (hunk 4 above)
Comment (jabraham17) [reply modules/packages/Python.chpl:365]: I think even better, its probably possible to just have the literal opaque extern type of `extern type PyObject`. I will look into that

File: modules/packages/Python.chpl [hunk 9]
```diff
@@ -0,0 +1,1417 @@
...
+    proc this(type retType, const args...): retType throws {
+      var pyArgs: args.size * c_ptr(void);
+      for param i in 0..#args.size {
+        pyArgs(i) = interpreter.toPython(args(i));
+      }
+      var pyRes;
+      if pyArgs.size == 1 then
+        pyRes = PyObject_CallOneArg(this.obj!.get(), pyArgs(0));
+      else
+        pyRes = PyObject_CallFunctionObjArgs(this.obj!.get(), (...pyArgs), nil);
```
[the remark this answers is not in the document]
Comment (jabraham17) [reply modules/packages/Python.chpl:981]: Yes, there is a lot of duplicated code right now between Function and ClassObject. I plan to introduce a class hierarchy to handle this in a future PR

File: modules/packages/Python.chpl (hunk 5 above)
Comment (jabraham17) [reply modules/packages/Python.chpl:392]: It cannot, since many of the cases rely on more than just simple equality between types.

File: modules/packages/Python.chpl [hunk 10]
```diff
@@ -0,0 +1,1417 @@
...
+
+     var interp = new Interpreter();
+     var func = new Function(interp, "lambda x,: x * x");
+
+     var ts = new threadState();
+     ts.save(); // save the thread state
+     coforall tid in 1..10 {
+       var gil = new GIL(); // acquire the GIL
+       Arr[tid] = func(int, tid);
+       gil.release(); // release the GIL
```
[the remark this answers is not in the document]
Comment (jabraham17) [reply modules/packages/Python.chpl:108]: The GIL/thread state as a context manager is future work. I put the explicit `.release` for clarity, but I can also just comment it

File: test/library/packages/Python/argPassingTest.chpl [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  catch e: PythonException { writeln("Caught PythonException: ", e.message()); }
+  catch                    { writeln("Caught unknown exception"); }
+}
+proc test_one_arg(mod: borrowed Module) {
+  const funcName = "one_arg";
+  var func = new Function(mod, funcName);
+
+  func(NoneType, 1);
+
+  // error: too many args
```
Comment (DanilaFe) [inline test/library/packages/Python/argPassingTest.chpl:26]: What about not enough args?

File: test/library/packages/Python/customType.chpl [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  override proc handlesType(type T): bool {
+    return T == myRec;
+  }
+  override proc toPython(interpreter: borrowed Interpreter,
+                         type T, const value: T): Python.CPythonInterface.PyObjectPtr throws {
+    if T != myRec then halt("Expected myRec");
+    var pyObj = owned.release(pyClsType(value.x, value.y));
+    pyObj.obj!.isOwned = false;
+    var ptr = pyObj.obj!.get();
+    delete pyObj;
```
Comment (DanilaFe) [inline test/library/packages/Python/customType.chpl:21]: Consider defining a `PyObject.release` type method that does this, so that you can do:

```Chapel
return PyObject.release(myClsType(value.x, value.y))
```

Since this is part of the user-facing API (that's how you expect users to define their own converters), I think dealing with the internals of `PyObject` and its `isOwned` field is not ideal.

**Review by DanilaFe — APPROVED** [review]
Your changes look good and I've read-skimmed the tests, trusting you on those.
